docs/AddressSpace.md

```diff
@@ -0,0 +1,153 @@
+14. The `StatsRange` makes its observations, the `GlobalRange` now unlocks the global component of the chain, and the `CommitRange` ensures that the allocation is mapped.
+    Aside from these side effects, these propagate the allocation along the chain unaltered.
+
+15. We now arrive back at the thread-local large buddy allocator, which takes its 2 MiB refill and breaks it down into powers of two down to the requested `MIN_CHUNK_SIZE`.
+    The second allocation (of the chunk), will either return or again break down one of these intermediate chunks.
+
+16. For the first (metadata) allocation, the thread-local *small* allocator breaks the `MIN_CHUNK_SIZE` allocation down into powers of two down to `PAGEMAP_METADATA_STRUCT_SIZE` and returns one of that size.
+    The second allocation will have been forwarded and so is not additionally handled here.
+
+Exciting, no?
+
+## What Can I Learn from the Pagemap?
+
+### Decoding a MetaEntry
+
+The centerpiece of `snmalloc`'s metadata is its `PageMap`, which associates each "chunk" of the address space (~16KiB; see `MIN_CHUNK_BITS`) with a `MetaEntry`.
+A `MetaEntry` is a pair of pointers, suggestively named `meta` and `remote_and_sizeclass`.
+In more detail, `MetaEntry`s are better represented by Sigma and Pi types, all packed into two pointer-sized words in ways that preserve pointer provenance on CHERI.
+
+To begin decoding, a bit (`REMOTE_BACKEND_MARKER`) in `remote_and_sizeclass` distinguishes chunks owned by frontend and backend allocators.
+
+For chunks owned by the *frontend* (`REMOTE_BACKEND_MARKER` not asserted),
+
+1. The `remote_and_sizeclass` field is a product of
+
+   1. A `RemoteAllocator*` indicating the `LocalAlloc` that owns the region of memory.
+
+   2. A "full sizeclass" value (itself a tagged sum type between large and small sizeclasses).
+
+2. The `meta` pointer is a bit-stuffed pair of
+
+   1. A pointer to a larger metadata structure with type dependent on the role of this chunk
+
+   2. A bit (`META_BOUNDARY_BIT`) that serves to limit chunk coalescing on platforms where that may not be possible, such as CHERI.
+
+See `src/backend/metatypes.h` and `src/mem/metaslab.h`.
+
+For chunks owned by a *backend* (`REMOTE_BACKEND_MARKER` asserted), there are again multiple possibilities.
+
+For chunks owned by a *small buddy allocator*, the remainder of the `MetaEntry` is zero.
+That is, it appears to have small sizeclass 0 and an implausible `RemoteAllocator*`.
+
+For chunks owned by a *large buddy allocator*, the `MetaEntry` is instead a node in a red-black tree of all such chunks.
+Its contents can be decoded as follows:
+
+1. The `meta` field's `META_BOUNDARY_BIT` is preserved, with the same meaning as in the frontend case, above.
+
+2. `meta` (resp. `remote_and_sizeclass`) includes a pointer to the left (resp. right) *chunk* of address space.
+   (The corresponding child *node* in this tree is found by taking the *address* of this chunk and looking up the `MetaEntry` in the Pagemap.)
```

Comment:
Why use the address of the chunk and not the `MetaEntry`? Wouldn't that make traversing the tree faster?

Reply:
If you use the address of the chunk, then it actually works better with CHERI.  We can get permission on the pagemap from a global, so we don't need a capability to access it.  But if you wanted direct pointers into the pagemap, either you would have to rederive them to follow, or you would need three capabilities, one for each direction, and one for the underlying allocation.  With pointers like this, we have access to the full capability for the chunks.  This makes the code almost unchanged for CHERI (@nwf spotted this, and saved me jumping through hoops for CHERI).

This representation adds a shift and add to each level of the tree, which given it is balanced and consolidated is not going to be many levels, over the direct pointers.  Also, this is not perf critical as it is many levels off the fast path.

Reply:
Thanks for calling that out; I'll add it to the documentation.

It's also probably worth noting that there's a *lot* of redundancy in the pagemap in CHERI at the moment: everyone, either in the Pagemap directly (as here) or in the `MetaCommon::chunk` for slabs is carrying around full `Arena` bounds to the entire underlying `mmap` region so that we can `madvise` and `Pal::zero` with `mmap` and, most challenging, recombine within the buddy allocators.  I'm not exactly thrilled with that, but I don't see an easy way to make anything smaller, since we do need pointers in the pagemap and so may as well use the space when we've got it, and while it might be good to more tightly bound internally (perhaps with yet more points in the `capptr::bounds::Spatial` taxonomy and maybe just in debug builds), it's not clear that there's a threat model where doing that actually distinguishes adversaries, since if an adversary can leak one high-authority pointer it can probably choose which one it gets.

Reply:
Makes sense, thanks. I was partly asking to check my understanding. I will keep @nwf's points in mind as I browse code.